packages/core/src/baseCogniteClient.ts

```diff
@@ -75,19 +75,21 @@ export default class BaseCogniteClient {
   private readonly metadata: MetadataMap;
   private readonly loginApi: LoginAPI;
   private readonly logoutApi: LogoutApi;
-  /**
-   * On 401 it might be possible to get a new token, but if `getToken` returns the same over and
-   * over (e.g api-keys) there isn't a point retrying. previousToken is used to keep track of that,
-   * comparing new tokens to one tried the last time.
-   */
-  private previousToken: string | undefined;
   private readonly getToken: () => Promise<string | undefined>;
   private readonly apiKeyMode: boolean;
   private readonly noAuthMode?: boolean;
   readonly project: string;
 
   private readonly credentialsAuth?: CredentialsAuth;
 
+  /**
+   * To prevent calling `getToken` method multiple times in parallel, we set
+   * the promise that `getToken` returns to a variable and check its existence
+   * on function calls for `authenticate`.
+   */
+  private tokenPromise?: Promise<string | undefined>;
+  private isTokenPromiseFulfilled?: boolean;
```

Comment:
Either we need to create our own promise class with statuses (I attempted sth like that for sdk-singleton long time ago: https://github.com/cognitedata/cdf-hub-sdk-singleton/pull/74/files#diff-39b2554fd18da165b59a6351b1aafff3714e2a80c1435f2de9706355b4d32351R6-R11) or we need to keep a boolean for that status. Right now we don't care about anything except `fulfilled` status - so this should be sufficient.